pkg/bootconfig/bootconfig.go

```diff
@@ -13,13 +20,49 @@ type BootConfig struct {
 	DeviceTree string `json:"devicetree,omitempty"`
 }
 
-// Validate returns true if a BootConfig object has valid content, and false
+// IsValid returns true if a BootConfig object has valid content, and false
 // otherwise
-func (bc *BootConfig) Validate() bool {
-	if bc.Kernel == "" {
-		return false
-	}
-	return true
+func (bc *BootConfig) IsValid() bool {
+    return bc.Kernel != ""
+}
+
+// Boot tries to boot the kernel with optional initramfs and command line
+// options. If a device-tree is specified, that will be used too
+func (bc *BootConfig) Boot() error {
+    // kexec: try the kexec executable first, and if it fails, use the native
+    // Go implementation of kexec from u-root
+    if err := kexecbin.KexecBin(bc.Kernel, bc.Initramfs, bc.KernelArgs, bc.DeviceTree); err != nil {
+        if os.IsNotExist(err) {
+            log.Printf("BootConfig: KexecBin failed, trying pure-Go kexec. Error: %v", err)
+        } else {
+            return err
+        }
+    }
+    kernel, err := os.Open(bc.Kernel)
+    if err != nil {
+        return err
+    }
+    var initramfs *os.File
+    if bc.Initramfs != "" {
+        initramfs, err = os.Open(bc.Kernel)
+        if err != nil {
+            return err
+        }
+    }
+    defer func() {
+        // clean up
+        if kernel != nil {
+            if err := kernel.Close(); err != nil {
+                log.Printf("Error closing kernel file descriptor: %v", err)
+            }
+        }
+        if initramfs != nil {
+            if err := initramfs.Close(); err != nil {
+                log.Printf("Error closing initramfs file descriptor: %v", err)
+            }
+        }
```

Comment:
Don't forget the devicetree

Reply:
it's not necessasry here - the devicetree field is a string, and it's only used with KexecBin. The cleanup is done there

Reply:
ah right, kexec file load does not use the dt. Sorry!